File: decoder/vt_decode.m
/* vt_decode.m — macOS VideoToolbox H.264→RGB decoder.
   Compile: clang -framework VideoToolbox -framework CoreMedia -framework CoreVideo -framework CoreFoundation -O2 -o vt_decode vt_decode.m
   Usage: ./vt_decode <input.264>
   Output: raw RGB24 frames to stdout, each preceded by a 4-byte frame size (big-endian).
*/

#import <VideoToolbox/VideoToolbox.h>
#import <CoreMedia/CoreMedia.h>
#import <CoreVideo/CoreVideo.h>
#import <CoreFoundation/CoreFoundation.h>
#import <stdio.h>
#import <stdlib.h>
#import <string.h>
#import <unistd.h>
#import <fcntl.h>

// Global state
static VTDecompressionSessionRef session = NULL;
static CMVideoFormatDescriptionRef fmtDesc = NULL;
static int width = 0, height = 0;
static FILE *out = NULL;

// Decode callback — called when VT produces a frame
static void decode_callback(void *refCon, void *srcFrameRef,
    OSStatus status, VTDecodeInfoFlags flags,
    CVImageBufferRef imageBuffer, CMTime pts, CMTime duration)
{
    if (status != noErr || !imageBuffer) return;

    CVPixelBufferLockBaseAddress(imageBuffer, kCVPixelBufferLock_ReadOnly);

    int w = (int)CVPixelBufferGetWidth(imageBuffer);
    int h = (int)CVPixelBufferGetHeight(imageBuffer);
    int stride = (int)CVPixelBufferGetBytesPerRow(imageBuffer);
    uint8_t *base = (uint8_t *)CVPixelBufferGetBaseAddress(imageBuffer);
    OSType fmt = CVPixelBufferGetPixelFormatType(imageBuffer);

    // Convert to RGB24
    uint8_t rgb[w * h * 3];
    for (int y = 0; y < h; y++) {
        for (int x = 0; x < w; x++) {
            uint8_t *pix = base + y * stride + x * 4;
            // BGRA → RGB
            rgb[(y * w + x) * 3 + 0] = pix[2];  // R
            rgb[(y * w + x) * 3 + 1] = pix[1];  // G
            rgb[(y * w + x) * 3 + 2] = pix[0];  // B
        }
    }

    CVPixelBufferUnlockBaseAddress(imageBuffer, kCVPixelBufferLock_ReadOnly);

    // Write frame size + data to stdout
    uint32_t sz = htonl(w * h * 3);
    fwrite(&sz, 4, 1, stdout);
    fwrite(rgb, w * h * 3, 1, stdout);
    fflush(stdout);
}

// Feed NAL unit to decoder
static void feed_nal(const uint8_t *nal, int len, int64_t pts_val) {
    if (!session || !nal || len == 0) return;

    // Create CMBlockBuffer
    CMBlockBufferRef blockBuf = NULL;
    OSStatus st = CMBlockBufferCreateWithMemoryBlock(
        kCFAllocatorDefault, (void *)nal, len, kCFAllocatorNull,
        NULL, 0, len, 0, &blockBuf);
    if (st != noErr) return;

    // Create CMSampleBuffer
    CMSampleBufferRef sampleBuf = NULL;
    CMSampleTimingInfo timing = {
        .duration = CMTimeMake(1, 24),
        .presentationTimeStamp = CMTimeMake(pts_val, 24),
        .decodeTimeStamp = kCMTimeInvalid
    };

    st = CMSampleBufferCreateReady(
        kCFAllocatorDefault, blockBuf, fmtDesc, 1, 1,
        &timing, 0, NULL, &sampleBuf);
    CFRelease(blockBuf);
    if (st != noErr) return;

    // Decode
    VTDecodeFrameFlags df = 0;
    VTDecodeInfoFlags info = 0;
    st = VTDecompressionSessionDecodeFrame(
        session, sampleBuf, df, NULL, &info);
    CFRelease(sampleBuf);
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "Usage: vt_decode <input.264>\n"); return 1; }

    // Open input file
    int fd = open(argv[1], O_RDONLY);
    if (fd < 0) { perror("open"); return 1; }

    // Read entire file (for growing files, we'll loop)
    uint8_t buf[16*1024*1024];  // 16MB max per pass
    ssize_t total = 0;
    while (1) {
        ssize_t n = read(fd, buf + total, sizeof(buf) - total - 1);
        if (n <= 0) break;
        total += n;
    }
    close(fd);

    if (total < 4) return 0;

    // Find SPS and PPS
    uint8_t *sps = NULL, *pps = NULL;
    int sps_len = 0, pps_len = 0;
    for (int i = 0; i < total - 4; i++) {
        if (buf[i] == 0 && buf[i+1] == 0 && buf[i+2] == 0 && buf[i+3] == 1) {
            int nal_type = buf[i+4] & 0x1F;
            int start = i + 4;
            int end = total;
            for (int j = start; j < total - 3; j++) {
                if (buf[j]==0 && buf[j+1]==0 && buf[j+2]==0 && buf[j+3]==1) {
                    end = j; break;
                }
            }
            if (nal_type == 7) { sps = buf + start; sps_len = end - start; }
            if (nal_type == 8) { pps = buf + start; pps_len = end - start; }
            i = end - 1;
        }
    }

    if (!sps || !pps) { fprintf(stderr, "No SPS/PPS found\n"); return 1; }

    // Create format description
    const uint8_t *param_ptrs[] = {sps, pps};
    size_t param_sizes[] = {sps_len, pps_len};
    OSStatus st = CMVideoFormatDescriptionCreateFromH264ParameterSets(
        kCFAllocatorDefault, 2, param_ptrs, param_sizes, 4, &fmtDesc);
    if (st != noErr) { fprintf(stderr, "CMVideoFormatDescription failed: %d\n", st); return 1; }

    // Get dimensions
    CMVideoDimensions dims = CMVideoFormatDescriptionGetDimensions(fmtDesc);
    width = dims.width; height = dims.height;

    // Create decompression session
    VTDecompressionOutputCallbackRecord cb = { decode_callback, NULL };

    // Pixel buffer attributes: request BGRA output
    CFNumberRef px_fmt = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type,
        (int[]){kCVPixelFormatType_32BGRA});
    CFNumberRef px_w = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &width);
    CFNumberRef px_h = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &height);
    const void *keys[] = {
        kCVPixelBufferPixelFormatTypeKey,
        kCVPixelBufferWidthKey,
        kCVPixelBufferHeightKey
    };
    const void *vals[] = {px_fmt, px_w, px_h};
    CFDictionaryRef px_attrs = CFDictionaryCreate(
        kCFAllocatorDefault, keys, vals, 3,
        &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);

    st = VTDecompressionSessionCreate(
        kCFAllocatorDefault, fmtDesc, NULL, px_attrs, &cb, &session);
    CFRelease(px_attrs); CFRelease(px_fmt); CFRelease(px_w); CFRelease(px_h);
    if (st != noErr) { fprintf(stderr, "VTDecompressionSessionCreate failed: %d\n", st); return 1; }

    // Feed all NAL units
    int64_t pts = 0;
    for (int i = 0; i < total - 4; i++) {
        if (buf[i]==0 && buf[i+1]==0 && buf[i+2]==0 && buf[i+3]==1) {
            int nal_start = i + 4;
            int nal_end = total;
            for (int j = nal_start; j < total - 3; j++) {
                if (buf[j]==0 && buf[j+1]==0 && buf[j+2]==0 && buf[j+3]==1) {
                    nal_end = j; break;
                }
            }
            int nal_type = buf[nal_start] & 0x1F;
            if (nal_type == 1 || nal_type == 5) {  // non-IDR or IDR
                // Feed accumulated NALs for this frame (SPS+PPS+slice)
                feed_nal(sps, sps_len, pts);
                feed_nal(pps, pps_len, pts);
                // Also feed preceding SEI if present
                feed_nal(buf + nal_start, nal_end - nal_start, pts);
                pts++;
            }
            i = nal_end - 1;
        }
    }

    // Flush
    VTDecompressionSessionWaitForAsynchronousFrames(session);
    VTDecompressionSessionFinishDelayedFrames(session);

    // Cleanup
    VTDecompressionSessionInvalidate(session);
    CFRelease(session);
    CFRelease(fmtDesc);

    return 0;
}

File: dash_qt/vt_decode.swift
import VideoToolbox; import CoreMedia; import Foundation

let args = CommandLine.arguments
guard args.count > 1 else { print("Usage: vt_decode <file.264>"); exit(1) }
let inPath = args[1]

// Global state
var session: VTDecompressionSession?
var fmtDesc: CMVideoFormatDescription?
var width: Int32 = 0; var height: Int32 = 0
var outFd = FileHandle.standardOutput
var frameCount: Int32 = 0

// Decode callback
func onFrame(_: UnsafeMutableRawPointer?, _: UnsafeMutableRawPointer?,
             status: OSStatus, _: VTDecodeInfoFlags,
             imageBuffer: CVImageBuffer?, _: CMTime, _: CMTime) {
    guard status == noErr, let buf = imageBuffer else { return }
    CVPixelBufferLockBaseAddress(buf, .readOnly)
    defer { CVPixelBufferUnlockBaseAddress(buf, .readOnly) }

    let w = CVPixelBufferGetWidth(buf)
    let h = CVPixelBufferGetHeight(buf)
    let stride = CVPixelBufferGetBytesPerRow(buf)
    let base = CVPixelBufferGetBaseAddress(buf)!.assumingMemoryBound(to: UInt8.self)

    var rgb = [UInt8](repeating: 0, count: w * h * 3)
    for y in 0..<h {
        for x in 0..<w {
            let src = base.advanced(by: y * stride + x * 4)
            let dst = (y * w + x) * 3
            rgb[dst] = src[2]; rgb[dst+1] = src[1]; rgb[dst+2] = src[0] // BGRA→RGB
        }
    }

    var sz = UInt32(w * h * 3).bigEndian
    withUnsafeBytes(of: &sz) { outFd.write(Data($0)) }
    outFd.write(Data(rgb))
    frameCount += 1
}

// Feed NAL unit to decoder
func feed(_ nal: [UInt8], _ pts: Int64) {
    guard let session = session, !nal.isEmpty else { return }
    var nalLen = UInt32(nal.count).bigEndian
    var nalData = Data()
    withUnsafeBytes(of: &nalLen) { nalData.append(Data($0)) }
    nalData.append(Data(nal))

    var blockBuf: CMBlockBuffer?
    nalData.withUnsafeBytes { ptr in
        CMBlockBufferCreateWithMemoryBlock(
            allocator: kCFAllocatorDefault, memoryBlock: nil,
            blockLength: nalData.count, blockAllocator: kCFAllocatorDefault,
            customBlockSource: nil, offsetToData: 0, dataLength: nalData.count,
            flags: 0, blockBufferOut: &blockBuf)
    }
    guard var bb = blockBuf else { return }
    defer { CFRelease(bb) }

    CMBlockBufferReplaceDataBytes(with: nalData, blockBuffer: &bb, offsetIntoDestination: 0, dataLength: nalData.count)

    var sampleBuf: CMSampleBuffer?
    var timing = CMSampleTimingInfo(
        duration: CMTime(value: 1, timescale: 24),
        presentationTimeStamp: CMTime(value: pts, timescale: 24),
        decodeTimeStamp: .invalid)
    CMSampleBufferCreateReady(
        allocator: kCFAllocatorDefault, dataBuffer: bb,
        formatDescription: fmtDesc, sampleCount: 1, sampleTimingEntryCount: 1,
        sampleTimingArray: &timing, sampleSizeEntryCount: 0,
        sampleSizeArray: nil, sampleBufferOut: &sampleBuf)
    guard var sb = sampleBuf else { return }
    defer { CFRelease(sb) }

    VTDecompressionSessionDecodeFrame(session, &sb, [], nil, nil)
}

// Main
func processFile(_ path: String) {
    guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)) else { return }
    let bytes = [UInt8](data); let n = bytes.count
    guard n > 4 else { return }

    // Find SPS/PPS (only once)
    if fmtDesc == nil {
        var sps: [UInt8] = [], pps: [UInt8] = []
        var i = 0
        while i < n - 3 {
            if bytes[i]==0 && bytes[i+1]==0 && bytes[i+2]==0 && bytes[i+3]==1 {
                let t = Int(bytes[i+4]) & 0x1F
                var end = n
                for j in (i+5)..<(n-3) {
                    if bytes[j]==0 && bytes[j+1]==0 && bytes[j+2]==0 && bytes[j+3]==1 { end = j; break }
                }
                if t == 7 { sps = Array(bytes[(i+4)..<end]) }
                if t == 8 { pps = Array(bytes[(i+4)..<end]) }
                i = end
            } else { i += 1 }
        }
        guard !sps.isEmpty, !pps.isEmpty else { return }

        sps.withUnsafeBufferPointer { sp in
            pps.withUnsafeBufferPointer { pp in
                var ptrs: [UnsafePointer<UInt8>] = [sp.baseAddress!, pp.baseAddress!]
                var sizes: [Int] = [sps.count, pps.count]
                CMVideoFormatDescriptionCreateFromH264ParameterSets(
                    allocator: kCFAllocatorDefault, parameterSetCount: 2,
                    parameterSetPointers: &ptrs, parameterSetSizes: &sizes,
                    nalUnitHeaderLength: 4, formatDescriptionOut: &fmtDesc)
            }
        }
        guard let desc = fmtDesc else { return }
        let dims = CMVideoFormatDescriptionGetDimensions(desc)
        width = dims.width; height = dims.height

        // Create session
        var cb = VTDecompressionOutputCallbackRecord(
            decompressionOutputCallback: onFrame, decompressionOutputRefCon: nil)
        var pxFmt = kCVPixelFormatType_32BGRA
        var w32 = width; var h32 = height
        var pxFmtNum = CFNumberCreate(kCFAllocatorDefault, .sInt32Type, &pxFmt)!
        var pxW = CFNumberCreate(kCFAllocatorDefault, .intType, &w32)!
        var pxH = CFNumberCreate(kCFAllocatorDefault, .intType, &h32)!
        defer { CFRelease(pxFmtNum); CFRelease(pxW); CFRelease(pxH) }
        let keys = [kCVPixelBufferPixelFormatTypeKey, kCVPixelBufferWidthKey, kCVPixelBufferHeightKey] as CFArray
        let vals = [pxFmtNum, pxW, pxH] as CFArray
        var pxAttrs = CFDictionaryCreate(kCFAllocatorDefault,
            unsafeBitCast(keys, to: UnsafePointer<UnsafeRawPointer?>.self),
            unsafeBitCast(vals, to: UnsafePointer<UnsafeRawPointer?>.self), 3,
            nil, nil)!
        defer { CFRelease(pxAttrs) }
        VTDecompressionSessionCreate(allocator: kCFAllocatorDefault,
            formatDescription: desc, videoDecoderSpecification: nil,
            destinationImageBufferAttributes: pxAttrs,
            outputCallback: &cb, decompressionSessionOut: &session)
    }

    guard let session = session, var sps: [UInt8] = nil else { return }
    // Rescan for SPS/PPS
    var sps2: [UInt8] = [], pps2: [UInt8] = []
    var i = 0
    while i < n - 3 {
        if bytes[i]==0 && bytes[i+1]==0 && bytes[i+2]==0 && bytes[i+3]==1 {
            let t = Int(bytes[i+4]) & 0x1F
            var end = n
            for j in (i+5)..<(n-3) {
                if bytes[j]==0 && bytes[j+1]==0 && bytes[j+2]==0 && bytes[j+3]==1 { end = j; break }
            }
            if t == 7 { sps2 = Array(bytes[(i+4)..<end]) }
            if t == 8 { pps2 = Array(bytes[(i+4)..<end]) }
            i = end
        } else { i += 1 }
    }
    guard !sps2.isEmpty, !pps2.isEmpty else { return }

    var pts: Int64 = frameCount
    i = 0
    while i < n - 3 {
        if bytes[i]==0 && bytes[i+1]==0 && bytes[i+2]==0 && bytes[i+3]==1 {
            let t = Int(bytes[i+4]) & 0x1F
            var end = n
            for j in (i+5)..<(n-3) {
                if bytes[j]==0 && bytes[j+1]==0 && bytes[j+2]==0 && bytes[j+3]==1 { end = j; break }
            }
            if t == 1 || t == 5 {
                feed(sps2, pts); feed(pps2, pts)
                feed(Array(bytes[(i+4)..<end]), pts)
                pts += 1
            }
            i = end
        } else { i += 1 }
    }
}

// Run: process file, wait for async frames
processFile(inPath)
if let s = session {
    VTDecompressionSessionWaitForAsynchronousFrames(s)
    VTDecompressionSessionFinishDelayedFrames(s)
    VTDecompressionSessionInvalidate(s)
}
